R/multivariate_karcher_mean.R

```diff
@@ -1,124 +1,165 @@
-multivariate_karcher_mean <- function (beta, lambda = 0.0, maxit = 20, ms = "mean")
+multivariate_karcher_mean <- function(beta,
+                                      rotation = FALSE,
+                                      scale = FALSE,
+                                      maxit = 20,
+                                      ms = c("mean", "median"))
 {
-  if(ms!="mean"&ms!="median"){warning("ms must be either \"mean\" or \"median\". ms has been set to \"mean\"",immediate. = T)}
-  if(ms!="median"){ms = "mean"}
+  ms <- rlang::arg_match(ms)
 
-  tmp = dim(beta)
-  n = tmp[1]
-  T1 = tmp[2]
-  time1 <- seq(0,1,length.out=T1)
-  N = tmp[3]
-  q = array(0, c(n, T1, N))
-  for (ii in 1:N) {
-    beta1 = beta[ , , ii]
-    out = curve_to_q(beta1, FALSE)
-    q[, , ii] = out$q
-  }
-
-  mu = q[, , 1]
-  bmu = beta[, , 1]
-  delta = 0.5
-  tolv = 1e-04
-  told = 5 * 0.001
-  itr = 1
-  sumd = rep(0, maxit + 1)
-  sumd[1] = Inf
-  betan = array(0, c(n, T1, N))
-  qn = array(0, c(n, T1, N))
-  normbar = rep(0, maxit + 1)
-  if(ms == "median"){ #run for median only, saves memory if getting mean
-    d_i = rep(0,N) #include vector for norm calculations
-    v_d = array(0, c(n, T1, N)) #include array to hold v_i / d_i
-  }
+  dims <- dim(beta)
+  L <- dims[1] # Dimension of codomain
+  M <- dims[2] # Size of the evaluation grid
+  N <- dims[3] # Sample size
+  q <- array(dim = c(L, M, N)) # Array for hosting SRVFs
 
+  preproc <- furrr::future_map(1:N, \(n) {
+    beta1 <- beta[ , , n]
+    out <- curve_to_q(beta1, scale = scale)
+    q1 <- out$q
+    if (scale)
+      beta1 <- beta1 / out$len
+    list(q1 = q1, beta1 = beta1)
+  }, .options = furrr::furrr_options(seed = TRUE))
 
-  cat("\nInitializing...\n")
-  gam = matrix(0,T1,N)
-  for (k in 1:N) {
-    out = find_rotation_seed_unqiue(mu, q[, , k], 'O', FALSE, FALSE, lambda)
-    gam[, k] = out$gambest
+  for (n in 1:N) {
+    q[ , , n] <- preproc[[n]]$q1
+    beta[ , , n] <- preproc[[n]]$beta1
   }
 
-  gam = t(gam)
-  gamI = SqrtMeanInverse(t(gam))
-  bmu = group_action_by_gamma_coord(bmu, gamI)
-  mu = curve_to_q(bmu, FALSE)$q
-  mu[is.nan(mu)] <- 0
+  # Initialize mean as the pointwise mean of the curves
+  # AST: improve with medoid, meaning compute distance matrix first.
+  qmean <- rowMeans(q, dims = 2)
+  betamean <- rowMeans(beta, dims = 2)
+  delta <- 0.5
+  tolv <- 1e-04
+  told <- 5 * 0.001
+  itr <- 1
+  sumd <- rep(0, maxit + 1)
+  sumd[1] <- Inf
+  betan <- betat <- array(0, c(L, M, N))
+  qn <- qt <- array(0, c(L, M, N))
+  normbar <- rep(0, maxit + 1)
+  if (ms == "median") {
+    # run for median only, saves memory if getting mean
+    d_i <- rep(0, N) #include vector for norm calculations
+    v_d <- array(0, c(L, M, N)) # include array to hold v_i / d_i
+  }
 
   while (itr < maxit) {
-    cat(sprintf("Iteration: %d\n", itr))
-    mu = mu
-
-    for (i in 1:N) {
-      q1 = q[, , i]
+    cli::cli_alert_info("Iteration {itr}/{maxit}...")
 
```

Comment:
After seeing this, this is very similar to `curve_karcher_mean` now, should we not just merge them and clean that one up?

Reply:
Good point too. I'll compare the two and try to see if there are any differences. If not, indeed, no need of both.